src/operator/join.rs

```diff
@@ -590,35 +611,67 @@ where
     fn clock_end(&mut self, scope: Scope) {
         debug_assert!(self
             .output_batchers
-            .iter()
-            .all(|(time, _)| !time.less_equal(&self.time)));
+            .keys()
+            .all(|time| !time.less_equal(&self.time)));
         self.time = self.time.advance(scope + 1);
     }
 
-    fn summary(&self, summary: &mut String) {
-        let sizes: Vec<usize> = self
+    fn metadata(&self, meta: &mut OperatorMeta) {
+        let total_size: usize = self
             .output_batchers
             .values()
             .map(|batcher| batcher.tuples())
-            .collect();
-        writeln!(summary, "sizes: {:?}", sizes).unwrap();
-        writeln!(summary, "total size: {}", sizes.iter().sum::<usize>()).unwrap();
+            .sum();
+
+        let batch_sizes = MetaItem::Array(
+            self.output_batchers
+                .values()
+                .map(|batcher| {
+                    let size = batcher.size_of();
+
+                    MetaItem::Map(
+                        metadata! {
+                            "allocated" => MetaItem::bytes(size.total_bytes()),
+                            "used" => MetaItem::bytes(size.used_bytes()),
+                        }
+                        .to_vec(),
+                    )
+                })
+                .collect(),
+        );
 
-        let mut context = Context::new();
-        for batcher in self.output_batchers.values() {
-            batcher.size_of_with_context(&mut context);
+        let bytes = {
+            let mut context = Context::new();
+            for batcher in self.output_batchers.values() {
+                batcher.size_of_with_context(&mut context);
+            }
+
+            context.size_of()
+        };
+
+        // Find the percentage of consolidated outputs
+        let mut output_redundancy = ((self.stats.output_tuples as f64
```

Comment:
Not sure this metric makes sense.  `produced_tuples` are tuples sent to the output stream of the operator. `output_tuples` are all computed outputs, including those stored inside `output_batchers`, which will be pushed to the output stream later.  It's not clear to me what this ratio represents.

Reply:
It's basically the ratio of produced outputs to consolidated outputs, if `(x, y)` is produced twice then `produced_tuples` will have a value of 2 while `output_tuples` will have a value of 1. It's basically a metric of "how many duplicate outputs are made from this join?"

Reply:
Yes, I figured this is what you were going for, but I don't think this formula reflects that due to the way join buffers outputs for future time stamps.

Reply:
Unrelated: can't this cause a divide-by-zero exception?

Reply:
No because of the magic of floats, it turns into nan or inf (hence the nan/inf check afterwards)

Reply:
Is that true in both debug and release builds?

Reply:
Yes